File: lib/telegram.rb
# Тонкая обёртка над Bot API на стандартной библиотеке.
# Без гемов сознательно: нужны четыре вызова, гем дороже.

require "net/http"
require "json"
require "uri"

module Telegram
  API_BASE = "https://api.telegram.org"

  # getFile отдаёт файлы не больше 20 МБ. Лимит платформы, обойти нельзя
  # (снимается только собственным Local Bot API Server).
  MAX_FILE_BYTES = 20 * 1024 * 1024

  # Лимиты Bot API. 4096 — текст сообщения, 1024 — подпись к документу.
  MAX_MESSAGE_CHARS = 4096
  MAX_CAPTION_CHARS = 1024

  class Error < StandardError; end

  module_function

  def token
    ENV["BOT_TOKEN"] or raise Error, "BOT_TOKEN не задан"
  end

  def http_for(uri, read_timeout: 60)
    http = Net::HTTP.new(uri.host, uri.port)
    http.use_ssl = true
    http.open_timeout = 10
    http.read_timeout = read_timeout
    http
  end

  def call(method, params, read_timeout: 60)
    uri = URI("#{API_BASE}/bot#{token}/#{method}")
    request = Net::HTTP::Post.new(uri, "Content-Type" => "application/json")
    request.body = JSON.generate(params)

    response = http_for(uri, read_timeout: read_timeout).request(request)
    parsed = JSON.parse(response.body)

    unless parsed["ok"]
      raise Error, "#{method}: #{parsed["description"] || response.code}"
    end

    parsed["result"]
  end

  def get_file(file_id)
    call("getFile", { file_id: file_id })
  end

  def download(file_path)
    uri = URI("#{API_BASE}/file/bot#{token}/#{file_path}")
    response = http_for(uri, read_timeout: 120).request(Net::HTTP::Get.new(uri))

    unless response.is_a?(Net::HTTPSuccess)
      raise Error, "скачивание файла: HTTP #{response.code}"
    end

    response.body
  end

  def send_message(chat_id, text, parse_mode: nil, reply_to: nil)
    params = { chat_id: chat_id, text: text }
    params[:parse_mode] = parse_mode if parse_mode
    params[:reply_parameters] = { message_id: reply_to } if reply_to
    call("sendMessage", params)
  end

  # Живёт примерно пять секунд. Для длинных операций переотправлять —
  # этим занимается Handle.with_typing.
  def send_chat_action(chat_id, action = "typing")
    call("sendChatAction", { chat_id: chat_id, action: action })
  end

  def send_document(chat_id, filename, content, caption: nil, reply_to: nil)
    uri = URI("#{API_BASE}/bot#{token}/sendDocument")
    boundary = "----tgbot#{Time.now.to_i}#{rand(1_000_000)}"

    fields = { "chat_id" => chat_id.to_s }
    fields["caption"] = caption if caption
    fields["reply_parameters"] = JSON.generate(message_id: reply_to) if reply_to

    body = +""
    fields.each do |name, value|
      body << "--#{boundary}\r\n"
      body << %(Content-Disposition: form-data; name="#{name}"\r\n\r\n)
      body << "#{value}\r\n"
    end
    body << "--#{boundary}\r\n"
    body << %(Content-Disposition: form-data; name="document"; filename="#{filename}"\r\n)
    body << "Content-Type: text/plain; charset=utf-8\r\n\r\n"
    body << content
    body << "\r\n--#{boundary}--\r\n"

    request = Net::HTTP::Post.new(uri)
    request["Content-Type"] = "multipart/form-data; boundary=#{boundary}"
    request.body = body.b

    response = http_for(uri, read_timeout: 120).request(request)
    parsed = JSON.parse(response.body)
    raise Error, "sendDocument: #{parsed["description"]}" unless parsed["ok"]

    parsed["result"]
  end

  # Экранирование для parse_mode=HTML.
  # Своя реализация вместо CGI.escapeHTML: не хочется зависеть от того,
  # останется ли cgi default-гемом в следующих версиях Ruby.
  def escape_html(text)
    text.to_s.gsub("&", "&amp;").gsub("<", "&lt;").gsub(">", "&gt;")
  end
end

File: lib/telegram.js
// Тонкая обёртка над Bot API. Только глобальный fetch, без зависимостей.

const API_BASE = "https://api.telegram.org";

// getFile отдаёт файлы не больше 20 МБ. Лимит платформы, снимается только
// собственным Local Bot API Server.
export const MAX_FILE_BYTES = 20 * 1024 * 1024;

// Лимиты Bot API: 4096 — текст сообщения, 1024 — подпись к документу.
export const MAX_MESSAGE_CHARS = 4096;
export const MAX_CAPTION_CHARS = 1024;

export class TelegramError extends Error {}

function token() {
  const value = process.env.BOT_TOKEN;
  if (!value) throw new TelegramError("BOT_TOKEN не задан");
  return value;
}

export async function call(method, params, { timeoutMs = 60_000 } = {}) {
  const response = await fetch(`${API_BASE}/bot${token()}/${method}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(params),
    signal: AbortSignal.timeout(timeoutMs),
  });

  const payload = await response.json();
  if (!payload.ok) {
    throw new TelegramError(`${method}: ${payload.description ?? response.status}`);
  }
  return payload.result;
}

export async function getFile(fileId) {
  return call("getFile", { file_id: fileId });
}

export async function download(filePath) {
  const response = await fetch(`${API_BASE}/file/bot${token()}/${filePath}`, {
    signal: AbortSignal.timeout(120_000),
  });

  if (!response.ok) {
    throw new TelegramError(`скачивание файла: HTTP ${response.status}`);
  }
  return Buffer.from(await response.arrayBuffer());
}

export async function sendMessage(chatId, text, { parseMode, replyTo } = {}) {
  const params = { chat_id: chatId, text };
  if (parseMode) params.parse_mode = parseMode;
  if (replyTo) params.reply_parameters = { message_id: replyTo };
  return call("sendMessage", params);
}

// Живёт около пяти секунд, поэтому на долгих операциях переотправляется —
// этим занимается withTyping в lib/handle.js.
export async function sendChatAction(chatId, action = "typing") {
  return call("sendChatAction", { chat_id: chatId, action });
}

export async function sendDocument(chatId, filename, content, { caption, replyTo } = {}) {
  const form = new FormData();
  form.set("chat_id", String(chatId));
  form.set("document", new Blob([content], { type: "text/plain" }), filename);
  if (caption) form.set("caption", caption);
  if (replyTo) form.set("reply_parameters", JSON.stringify({ message_id: replyTo }));

  const response = await fetch(`${API_BASE}/bot${token()}/sendDocument`, {
    method: "POST",
    body: form,
    signal: AbortSignal.timeout(120_000),
  });

  const payload = await response.json();
  if (!payload.ok) {
    throw new TelegramError(`sendDocument: ${payload.description ?? response.status}`);
  }
  return payload.result;
}

// Экранирование для parse_mode=HTML. Три замены, библиотека не нужна.
export function escapeHtml(text) {
  return String(text ?? "")
    .replaceAll("&", "&amp;")
    .replaceAll("<", "&lt;")
    .replaceAll(">", "&gt;");
}
